xray/commands/curation/audit.go

```diff
@@ -0,0 +1,504 @@
+	Condition string `json:"condition"`
+}
+
+type PackageStatusTable struct {
+	Status            string        `col-name:"Action"`
+	ParentName        string        `col-name:"Direct Dependency\nPackage Name"`
+	ParentVersion     string        `col-name:"Direct Dependency\nPackage Version"`
+	BlockedPackageUrl string        `col-name:"Blocked Package URL"`
+	PackageName       string        `col-name:"Blocked Package\nName"`
+	PackageVersion    string        `col-name:"Blocked Package\nVersion"`
+	BlockingReason    string        `col-name:"Blocking Reason"`
+	PkgType           string        `col-name:"Package Type"`
+	Policy            []policyTable `embed-table:"true"`
+}
+
+type policyTable struct {
+	Policy    string `col-name:"Violated Policy\nName"`
+	Condition string `col-name:"Violated Condition\nName"`
+}
+
+type treeAnalyzer struct {
+	rtManager            artifactory.ArtifactoryServicesManager
+	extractPoliciesRegex *regexp.Regexp
+	rtAuth               auth.ServiceDetails
+	httpClientDetails    httputils.HttpClientDetails
+	url                  string
+	repo                 string
+	tech                 coreutils.Technology
+	parallelRequests     int
+}
+
+type CurationAuditCommand struct {
+	PackageManagerConfig *rtUtils.RepositoryConfig
+	extractPoliciesRegex *regexp.Regexp
+	workingDirs          []string
+	OriginPath           string
+	parallelRequests     int
+	*utils.GraphBasicParams
+}
+
+func NewCurationAuditCommand() *CurationAuditCommand {
+	return &CurationAuditCommand{
+		extractPoliciesRegex: regexp.MustCompile(extractPoliciesRegexTemplate),
+		GraphBasicParams:     &utils.GraphBasicParams{},
+	}
+}
+
+func (ca *CurationAuditCommand) setPackageManagerConfig(pkgMangerConfig *rtUtils.RepositoryConfig) *CurationAuditCommand {
+	ca.PackageManagerConfig = pkgMangerConfig
+	return ca
+}
+
+func (ca *CurationAuditCommand) SetWorkingDirs(dirs []string) *CurationAuditCommand {
+	ca.workingDirs = dirs
+	return ca
+}
+
+func (ca *CurationAuditCommand) SetParallelRequests(threads int) *CurationAuditCommand {
+	ca.parallelRequests = threads
+	return ca
+}
+
+func (ca *CurationAuditCommand) Run() (err error) {
+	rootDir, err := os.Getwd()
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+	if len(ca.workingDirs) > 0 {
+		defer func() {
+			err = errorutils.CheckError(os.Chdir(rootDir))
+		}()
+	} else {
+		ca.workingDirs = append(ca.workingDirs, rootDir)
+	}
+	results := map[string][]*PackageStatus{}
+	for _, workDir := range ca.workingDirs {
+		var absWd string
+		absWd, err = filepath.Abs(workDir)
+		if err != nil {
+			return errorutils.CheckError(err)
+		}
+		log.Info("Running curation audit on project:", absWd)
+		if absWd != rootDir {
+			if err = os.Chdir(absWd); err != nil {
+				return errorutils.CheckError(err)
+			}
+		}
+		// If error returned, continue to print results(if any), and return error at the end.
+		err = ca.doCurateAudit(results)
+	}
+	if ca.Progress != nil {
+		if err = ca.Progress.Quit(); err != nil {
```

Comment:
When we get to this point, we may have an error stored in `err`. This line can override this err by nil.